### PR TITLE
Feature/base s3 publisher

### PR DESCRIPTION
**Pull request by akshayn1107**
NOTE: Do not merge until we merge the SimpleDataWriter - I just want to get this out so the code reviews could happen in parallel. This relies on some of the refactors that happened in the SimpleDataWriter (discussed with LinkedIn).

Anyways, this is the code for the S3 publisher. I've abstracted it out into a base class and a Kafka specific class for our needs (where we can define the bucket/key however we want).

Code Review: @vikarmic @eogren

### CODE REVIEW

File: gobblin-core/src/main/java/gobblin/publisher/BaseS3Publisher.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    for (WorkUnitState task : tasks) {
+      for (int i = 0; i < numBranches; i++) {
+        BucketAndKey bk = getBucketAndKey(task, i);
+        Long contentLength = contentLengths.get(bk);
+        if (contentLength == null) {
+          contentLength = 0L;
+        }
+        if (contentLength >= MAX_S3_FILE_SIZE) {
+          LOG.info(String.format("Content length <" + contentLength + "> exceeds max size <" + MAX_S3_FILE_SIZE + ">"));
+          InputStream is = data.get(bk);
```
Comment (vikarmic) [inline gobblin-core/src/main/java/gobblin/publisher/BaseS3Publisher.java]: Not sure what this is doing; is it a refactor artifact?

**Comment by vikarmic** [comment]
I think this is pretty solid; add some comments as discussed offline and lgtm.

**Comment by akshayn1107** [comment]
Integration tests pass. Once you guys have a chance to code review it, we can merge it in and load test this in stage :)

**Comment by vikarmic** [comment]
Awesome.

**Comment by vikarmic** [comment]
Looks good, nice refactor.

**Pull request merged**